### PR TITLE
[bugfix] Causal mask being skipped if no extra mask

### LINKED ISSUE #103: [bug] Causal mask is skipped if no att_mask is passed
# 🐛 Bug

In https://github.com/facebookresearch/xformers/blob/main/xformers/components/attention/scaled_dot_product.py#L78
if no att_mask is passed, then an hypothetical causal mask is completely skipped

### PR DESCRIPTION
**Pull request by blefaudeux**
## What does this PR do?
- Fixes #103
- Adding a unit test checking that causality is correct for scaled dot product and FAVOR. Nystrom does not seem to pass that
- Fixes Favor not picking up the causal flag
- Fixes Favor causal not being trainable (but introduces a big memory use)

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [x] Did you make sure to update the docs?
  - [ ] N/A
- [x] Did you write any new necessary tests?
  - [ ] N/A
- [x] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Review by SeanNaren — APPROVED** [review]

**Comment by blefaudeux** [comment]
The test I came up with is a bit strict, so it's effective with scaled_dot_product but it does not pass with all the approximations, they're mostly there but not exactly. I think it would be nice to come up with a proper test for all of them, but in the meantime this needs to be fixed on main

File: xformers/components/attention/favor.py [hunk 1]
```diff
@@ -24,6 +24,7 @@
 
 @dataclass
 class FavorAttentionConfig(AttentionConfig):
+    causal: Optional[bool]
```
Comment (blefaudeux) [inline xformers/components/attention/favor.py:27]: causal was not being passed on with the factory, because of that..

**Comment by blefaudeux** [comment]
back to draft, looks like favor causal needs improvements, long overdue (spotted by @fmassa long ago)

File: xformers/components/attention/favor.py [hunk 2]
```diff
@@ -109,33 +110,28 @@ def _maybe_promote(x: torch.Tensor) -> torch.Tensor:
     def _causal_attention(
         k_prime: torch.Tensor, q_prime: torch.Tensor, v: torch.Tensor
     ) -> Tuple[torch.Tensor, torch.Tensor]:
-        # TODO(@lefaudeux): Rewrite as suggested in
-        # https://github.com/fairinternal/xformers/pull/61#discussion_r628420093
         # Algorithm 1 in the paper
-        ref_v = torch.ones_like(v[:, 0, :].unsqueeze(1))
+        ref_v = torch.ones_like(v.unsqueeze(2))  # BATCH x SEQ x 1 x EMB
```
Comment (blefaudeux) [inline xformers/components/attention/favor.py:114]: this is a memory hog, that's the issue with this approach @fmassa, but the older state was not actually trainable (this was not being tested.. fixed now), so I guess that it's better than before and improving on that could be a follow up

Comment (blefaudeux) [reply xformers/components/attention/favor.py:114]: see https://github.com/facebookresearch/xformers/issues/105 for tracking

**Comment by blefaudeux** [comment]
@dianaml0 breaking now around Nystrom + key padding mask, I'm seeing a few lines that I don´t get (like mask and key_padding_mask being used interchangeably ?)

**Comment by dianaml0** [comment]
> @dianaml0 breaking now around Nystrom + key padding mask, I'm seeing a few lines that I don´t get (like mask and key_padding_mask being used interchangeably ?)

Taking a look now! For Nystrom only `key_padding_mask` is accepted. And if the causal flag is specified, `key_padding_mask` is combined with a causal mask into `mask`. I'll check the errors

File: xformers/components/attention/nystrom.py [hunk 3]
```diff
@@ -194,8 +194,9 @@ def forward(
             mask: Optional[torch.Tensor] = None
 
             if self.causal:
-                mask = self._tril_mask(batched_dim, seq_len, seq_len, **tt)
+                mask = self._triu_mask(batched_dim, seq_len, seq_len, **tt)
 
+            # FIXME: @lefaudeux This is probably broken, we don´t test this I suppose
```
Comment (dianaml0) [inline xformers/components/attention/nystrom.py:199]: This should be tested in `test_masking` in `test_nystrom_attention.py`

File: tests/test_nystrom_attention.py [hunk 4]
```diff
@@ -85,8 +85,10 @@ def test_att_mask_ignored():
         assert torch.allclose(r_nystrom, r_sdp, rtol=0.005, atol=1e-2)
 
     def test_masking():
-        nystrom_config["causal"] = True
-        sdp_config["causal"] = True
+        # FIXME: Masking seems fine as such, but whole lines are randomly masked when causal+random mask
```
Comment (dianaml0) [inline tests/test_nystrom_attention.py:88]: Is this what you meant by key padding + causal being broken?

File: xformers/components/attention/nystrom.py (hunk 3 above)
Comment (blefaudeux) [reply xformers/components/attention/nystrom.py:199]: it's this line which I thought could not possibly work, given the difference in shapes ?

File: tests/test_nystrom_attention.py (hunk 4 above)
Comment (blefaudeux) [reply tests/test_nystrom_attention.py:88]: yes, although it feels like it could be only an issue with the unit test, but scaled_dot_product and nystrom are not behaving in the same way (although there are plenty of NaNs around in both cases). I think that it could be that Nystrom breaks earlier because of the pooling for anchors, if one anchor is full of inf (easy with causal) it cannot compute the attention

Comment (blefaudeux) [reply tests/test_nystrom_attention.py:88]: Fixed now with a random masking a little less aggressive (not masking half of the words :)), I think that it's still a good test

File: xformers/components/attention/nystrom.py (hunk 3 above)
Comment (dianaml0) [reply xformers/components/attention/nystrom.py:199]: I think it should work since we assert on `key_padding_mask` dimension being (batch, 1, seq)? And we also convert it to an additive mask. But perhaps using `masked_fill` with a boolean key padding mask makes more sense?

File: xformers/components/attention/scaled_dot_product.py [hunk 5]
```diff
@@ -74,24 +74,40 @@ def forward(
                     - If the mask has the float type, then an additive mask is expected (masked values are -inf)
 
         """
+
+        # Handle a possibly deferred causal mask handling
```
Comment (dianaml0) [inline xformers/components/attention/scaled_dot_product.py:78]: Nice catching of edge cases!

**Review by dianaml0 — APPROVED** [review]

File: tests/test_nystrom_attention.py (hunk 4 above)
Comment (dianaml0) [reply tests/test_nystrom_attention.py:88]: Nice :) So can causal be used for the test now or that's still an issue?

File: tests/test_attentions.py [hunk 6]
```diff
@@ -81,7 +96,9 @@ def test_order_invariance(
     )
 
     # Check that a shuffled input produces the same results
-    seqs = [SEQ, SEQ - 16] if attention_name != "blocksparse" else [SEQ]
+    seqs = (
+        [SEQ, SEQ - 16] if (attention_name != "blocksparse" and not causal) else [SEQ]
```
[the remark this answers is not in the document]
Comment (blefaudeux) [reply tests/test_attentions.py:100]: ah, good question ! Basically we adjust the mask on the fly when the sequence is smaller than initially thought, but we don't have the codepath in place to also adjust the causal mask so that broke (it "worked" before because we were not actually applying the causal mask..). I can fix that in this PR and then refactor all this part in a sequel ?

File: tests/test_nystrom_attention.py (hunk 4 above)
Comment (blefaudeux) [reply tests/test_nystrom_attention.py:88]: not an issue anymore, I messed up the push when rebasing but it works fine locally, let me update..

File: xformers/components/attention/nystrom.py (hunk 3 above)
Comment (blefaudeux) [reply xformers/components/attention/nystrom.py:199]: yes you're right, sorry I forgot again, the 1 will be enough to broadcast ! all good, 'll remove the FIXME, I've changed the unit test to make sure that we do go there and it works indeed

File: tests/test_attentions.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dianaml0) [reply tests/test_attentions.py:100]: Ah okay thanks for the explanation :) That sounds good

[the remark this answers is not in the document]
Comment (dianaml0) [reply tests/test_attentions.py:100]: Ah okay thanks for the explanation :) That sounds good
